storm-core/src/clj/backtype/storm/daemon/executor.clj

```diff
@@ -445,6 +445,32 @@
     ret
     ))
 
+(defn init-batching-buffer [worker-context component-id]
+  (let [batchSize (.get_batch_size (.getComponentCommon worker-context component-id))]
+    (if (> batchSize 1)
+      (let [consumer-ids (flatten (for [cids (vals (.getTargets worker-context component-id))] (keys cids)))
+            consumer-task-ids (flatten (for [cid consumer-ids :when (not (.startsWith cid "__"))] (into '() (.getComponentTasks worker-context cid))))]
+        (HashMap. (zipmap consumer-task-ids (repeatedly (count consumer-task-ids) #(Batch. batchSize)))))
+      (HashMap.)
+      )))
+
+(defn emit-msg [out-task out-tuple overflow-buffer output-batch-buffer transfer-fn]
```

Comment:
Try type-hinting `output-batch-buffer` here. I haven't profiled it myself, but what you're saying as far as performance goes sounds like it could be the result of reflection. Type-hinting anything you're calling methods on should help.